### ISSUE TITLE
Building PH Mastodon Bot

### ISSUE THREAD

**Issue by ZoeLeBlanc** [programminghistorian/jekyll#2817]
Starting a new issue to discuss building a mastodon bot to advertise PH lessons.

Thanks to @GeeAlmeida for this tutorial https://shkspr.mobi/blog/2018/08/easy-guide-to-building-mastodon-bots/

A few initial thoughts:
- We would likely still want to use the google sheets interface for creating toots
- We need the bot to be somewhat a cron job (as in we don't want to have to run it manually)
-  Ideally it would push from Mastodon to Twitter too

Any other thoughts or features @jrladd @GeeAlmeida @anisa-hawes ?

**Comment by jrladd** [comment]
Great! So I think the high-level overview is a python script that does the following:

1. Reads the toots/tweets from the google sheet via the [Google Workspace API](https://developers.google.com/sheets/api/quickstart/python) or something similar.
2. Posts to Mastodon via the [Mastodon API](https://github.com/halcy/Mastodon.py).
3. Runs on its own automatically, via a cron job or similar mechanism.

Steps 1 & 2 can be repurposed from [the OG bot](https://github.com/programminghistorian/proghistbot/blob/master/bot.py). The secret fourth step is that the Mastodon account then cross-posts to Twitter, but that won't need to be part of the script itself.

A couple initial questions:

- It makes sense to work directly with the original bot code. Would folks prefer that we fork that repo or work on it directly? (If directly, I'd need admin access to the repo.)
- Hosting is still an issue, directly related to #2694. We could keep running the bot on Heroku or try to find an alternative hosting solution. Something like this could even run on a small local machine like a Raspberry Pi, but then somebody would have to be responsible for that machine. [Botwiki](https://botwiki.org/resources/hosting-platforms/) lists a bunch of possible alternatives.

**Comment by GeeAlmeida** [comment]
Thank you for asking @ZoeLeBlanc! 
I would like to be able to pause the bot if/when needed.
We need to be able to add/ edit the content of the posts (obvious, but thought I might add).

At the moment everything we post to Twitter shows up in Mastodon, and vice-versa. I registered with https://moaparty.com/ for this. However, Mastodon has a category of profiles exclusive for bots. I don't know if this affects what we are trying to do.

**Comment by GeeAlmeida** [comment]
I just want to add that I was wrong, at present what we publish on Twitter shows in Mastodon, but what is published on Mastodon does not show up on Twitter. I tested today (13/01).

**Comment by anisa-hawes** [comment]
Hello @GeeAlmeida. I came across a conversation recently, in which I learned that it is considered poor etiquette in the fediverse to cross-post from Twitter to Mastodon. I can't find the thread I read now, but [this post](https://mstdn.social/@feditips/107882830540203871) explains something similar. Perhaps we could aim to cross-post in the other direction? Just a thought.

**Comment by jrladd** [comment]
@GeeAlmeida It may still be possible to configure Moa Party to crosspost from Mastodon to Twitter. I've heard the same thing that @anisa-hawes has—that crossposting from Twitter to Mastodon is sometimes considered impolite. But since we want to originate posts on Mastodon with the bot, we can focus on the Mastodon-to-Twitter direction.

If Moa Party won't work, there are other services that will do this. As a potentially more reliable alternative to a standalone service, every Mastodon account has an RSS feed, and you can [use IFTTT to link an RSS feed to Twitter](https://edtechfactotum.com/cross-posting-from-mastodon-to-twitter-using-ifttt-and-rss/). The good news is that all of the crossposting is external to the bot itself, so it's easy to make changes to how we want to crosspost.

**Comment by GeeAlmeida** [comment]
@jrladd I did not understand everything you say, but that's ok lol. I agree that the focus should be on M --> T.

**Comment by GeeAlmeida** [comment]
@jrladd let me know if there is any information you need from me?

**Comment by jrladd** [comment]
@GeeAlmeida The only thing I can think of right now is: do you already have a Mastodon account that you want to run the bot on? Or should we create a new account? One thing I'll need to set up the bot is the username and password for the account we're using. (Also botsin.space is the Mastodon instance most commonly used for bots, so if we're making a new account we could use that instance.)

**Comment by anisa-hawes** [comment]
Hello @jrladd. Yes, @GeeAlmeida created one for us: 
[@proghist@hcommons.social](https://hcommons.social/@proghist).

- [x] I can share the username and password with you.

**Comment by jrladd** [comment]
The new bot is up and running for testing at [jrladd/proghistbot](https://github.com/jrladd/proghistbot). I think we should wait a couple days to make sure it's working properly, then we can move the code to the original bot repo and close this issue.

**Comment by anisa-hawes** [comment]
Hello @jrladd. 

I hope you're well?

I'm not working today, so I don't have time to take a closer look at the PR you've just merged in but ...

I have noticed that all the Tweets posted since Thursday (which I understand originate from the new bot) exclude the @ symbols that create links to any mentioned account(s). Is this a choice made due to the migration from Twitter to Mastodon? How do 'mentions' work in Mastodon? 

Many of the Tweets in our original Google Sheet include the account handles of contributors, so it would be good to maintain that (on Twitter at least, if we can).

Thank you for all the energy you've given to this already.

**Comment by jrladd** [comment]
Hi @anisa-hawes, thanks for bringing this to my attention! All of the Mastodon posts, like [this one from a few days ago](https://hcommons.social/@proghist/109723251500393338) include the @ symbol that was in the text from the spreadsheet. And you're right that the [corresponding tweet](https://twitter.com/ProgHist/status/1616523925554331649), forwarded from Mastodon, is missing this symbol.

Since this is isn't happening when posts first appear on Mastodon, I suspect it has something to do with how posts are processed when they're cross-posted to Twitter. @GeeAlmeida do you know what could be causing this?

On Mastodon, mentions work the same way they do on Twitter, just with the longer Mastodon handles (e.g. `@proghist@hcommons.social`). A Twitter handle on Mastodon will never automatically link to that user on Twitter, but a crossposted Twitter handle with an @ symbol should still link correctly. If future spreadsheet entries included Mastodon handles, those would link on the Mastodon posts. I hope that helps!

**Comment by anisa-hawes** [comment]
Hmmm... Thank you for your reply, @jrladd. I wonder how we can tackle this... 

Is there a way we could prepare two variations of each post, one including Mastodon handles (where available) and one including Twitter handles, then program the bot to select which to post where?

Or perhaps this is a sticking point, which asks us to reconsider whether cross-posting is actually the right solution for us in this context?

I understand the benefits that automation provides us, but this issue prompts me to think again about whether we could/should approach use of the two platforms independently.

Upon my initial explorations of Mastodon, I did feel the register and etiquette of interactions are quite different than Twitter.

I would be interested to hear your (@jrladd), @GeeAlmeida's and others' thoughts.

**Comment by jrladd** [comment]
I agree that Twitter and Mastodon have somewhat different registers, but I don't think PH's use of Twitter is different enough from the general vibe of Mastodon to require a different set of posts.

I think the Twitter handle issue is solvable on the technical side. I found this [open Moa Party PR](https://gitlab.com/fedstoa/moa/-/merge_requests/10) that shows they're working on similar issues with usernames. I suspect there's a problem with the way usernames are "sanitized" when passed from one site to another. This might be a setting we can change, or if Moa Party won't work for us we might try a different crossposting service. 

(We might also adopt a different convention for how we refer to Twitter usernames in our spreadsheet, that would then be correctly converted, e.g. `@johnrladd@twitter.com`. But before we change things on our end we should see if a crossposting tool can handle what we already have.)

If those options won't work, it's possible to program the bot to post to both Twitter and Mastodon, as you outlined. But I think that adds a complexity to the bot code/Github Action that we don't really need, and it would make the writing of tweets in the spreadsheet a little more complicated for editors. I think we should avoid this option if we can get the crossposting to work in a way that makes sense for us.

**Comment by anisa-hawes** [comment]
Thank you for reopening this Issue, @jrladd. I think it is important that we aim to preserve the links to Twitter handles if we can.

I think we need to consider working towards a system that allows us to create meaningful links to existing profiles in Twitter, as well new profiles in Mastodon. 

If the preparation of our Bot's posts requires one extra step, I think we can manage that. When working with new editors recently, I have been co-ordinating to add their Tweets to the spreadsheet and I could take this on across all lessons at the point of typesetting and final pre-publication steps if need be.

**Comment by acrymble** [comment]
@jrladd Thanks for looking into this further. Can I be a devils advocate and ask if this is worth extra time on your part to fix? It seems to me like the bot simply means we have to post a social media text once instead of twice. The cost savings of that is quite small and we have a dedicated employee whose responsibilities include that posting time. We don't always need technical solutions, I suppose.

**Comment by GeeAlmeida** [comment]
> could prepare two variations of each post, one including Mastodon handles (where available) and one including Twitter handles, then program the bot to select which to post where?

There's also the issue that Mastodon allows for more characters. Including the longer Mastodon handles is bound to make the text quite long for Twitter's 280 characters. I don't know what happens if a message is ok for Mastodon (max. 500 characters) but too long for Twitter. Maybe Moa Party splits into two Tweets?

**Comment by GeeAlmeida** [comment]
@acrymble I am not sure what you are suggesting in your comment. The bot is there precisely so that posting is automated - so it doesn't take any of my time. I spend part of my time sourcing and sharing other content. I think that works quite well, because I can identify CFPs and opportunities of interest to our community. The bot tweets the "same" all the time - links to our lessons.

**Comment by GeeAlmeida** [comment]
@jrladd I am confused with the way Mastodon posts are getting replicated on Twitter. I published something 1 hr ago on Mastodon that is not yet on Twitter (updated French call for editors). I published something later on Mastodon, 25 minutes ago, and that is already on Twitter (link to YouTube Portuguese event).
Does anyone have any idea of what might be going on?

**Comment by GeeAlmeida** [comment]
An update: I have tried something different, a scheduling tool for Mastodon (https://www.scheduler.mastodon.tools/)

The post has appeared in Mastodon, but not on Twitter.

That's not good, because scheduling is quite good, given that I am only here two days a week.

**Comment by GeeAlmeida** [comment]
In the previous bot each pair of tweets had to have a link. Is that still the case? Are links mandatory?

**Comment by anisa-hawes** [comment]
Thank you, @GeeAlmeida. Links are not mandatory, but they are useful to encourage action from people reading our posts!

**Comment by GeeAlmeida** [comment]
@programminghistorian/project-team 

I have created new tweets for the new communication tab of our new bot (I know, everything is new :)) 

I used an online translator to have the text across all 4 languages. I would really appreciate it if people could have a look and edit the following document as required:

https://docs.google.com/spreadsheets/d/1waG6N05X63PC8nlXgFb0tfxnXOpNKHyUmuvNg-yj7pM/edit?usp=sharing 

Please note that the messages are in different tabs (EN, ES, PT and FR).

Let me know below once you've had a chance to have a look / what language you have checked.

Thanks!

**Comment by jrladd** [comment]
Thanks @acrymble @anisa-hawes @GeeAlmeida for all of this. In general, I do think it's worthwhile to have some automated posts. To clarify, here's the workflow as it stands:

1. All the posts are entered via the Google Sheet. We can adjust these however we want (to include/remove links, change usernames, etc).
2. The bot takes whatever's in the Google Sheet, exactly as typed, and posts it to Mastodon. It doesn't make any adjustments to the posts, and the only substantive change I've made to the bot is that it posts to Mastodon instead of Twitter.
3. The third-party service [Moa](https://moa.party/) reposts whatever appears on Mastodon to Twitter. @GeeAlmeida set this up, and we don't have a lot of control over how it works. It seems to be removing the @ symbols from the usernames, and we might solve this by entering the usernames in our spreadsheet as `@johnrladd@twitter.com`. We could test this to see if the behavior works as expected. 

There are a number of other services that also crosspost between Mastodon and Twitter, and if Moa is giving us issues, it may be worth switching. But none of that really affects the operation of the bot. The same is true about the timing of when posts appear on Twitter from Mastodon and vice versa: all of that is handled by Moa, and I'm not really sure how/when things are crossposted.

I'll add that given recent events at Twitter I think it's good that we're using a Mastodon-first social media strategy, and while we definitely want to fix any bugs with how Twitter crossposting is currently working, it's probably wise that we aren't as singularly invested in that service. I hope that helps!

**Comment by marie-flesch** [comment]
I've edited the French tweets @GeeAlmeida

**Comment by DanielAlvesLABDH** [comment]
@GeeAlmeida just edited the PT version. Thanks for this

**Comment by GeeAlmeida** [comment]
> 3\. The third-party service [Moa](https://moa.party/) reposts whatever appears on Mastodon to Twitter. @GeeAlmeida set this up, and we don't have a lot of control over how it works. It seems to be removing the @ symbols from the usernames, and we might solve this by entering the usernames in our spreadsheet as `@johnrladd@twitter.com`. We could test this to see if the behavior works as expected.

Thanks for all your help with this issue @jrladd - Following this conversation I have communicated to Moa Party our issue. Let's see if they can help us fix this bug.

**Comment by GeeAlmeida** [comment]
Thanks you @marie-flesch and @DanielAlvesLABDH. The French and Portuguese have been checked.

@anisa-hawes would you mind reading through the posts in English? They are here: https://docs.google.com/spreadsheets/d/1waG6N05X63PC8nlXgFb0tfxnXOpNKHyUmuvNg-yj7pM/edit?usp=sharing

@programminghistorian/spanish-team could someone go through the posts in Spanish in the link above?

Thanks everyone, your help is much appreciated!

**Comment by rivaquiroga** [comment]
@GeeAlmeida, the Spanish versions are now checked.

I share all the concerns about Twitter, but fixing the bugs related to how our tweets are been published seems to me like a critical issue: 

1. Mentioning authors and people involved in a lesson publication is the way we have to publicly acknowledge their work. If they don’t receive a notification, there is no way they’ll know we are recognizing the time they invested in our project. 
2. Usually, the people involved in the lesson are the ones who retweet the message, and that is the way we have to reach new audiences, especially people that are not aware of our existence. 
3. Mentions are clickable: someone can go and see that person's profile and learn more about their work/research/projects. Our account on Twitter has +13k followers, so for some people, it might be an opportunity to make their work more visible to the DH community. 
4. Not everyone has a username that matches their name/surname, so not always is possible to infer the name of the person. 

The reason we have a Mastodon account is that there are DH people there and we want to keep connected with them. But we must take into account that the DH community in Mastodon is currently very biased towards people in North America and Europe. And that is just a fraction of our community.

**Comment by GeeAlmeida** [comment]
Hi @rivaquiroga Riva thanks for this. I agree we need to fix this, I just don't know how.
I am currently waiting on a reply from the Moa Party team. They are on Git Hub, and I started an issue in their repository.

Do you have a suggestion of something else we could do to solve this issue?

And thanks for checking the Spanish!

**Comment by acrymble** [comment]
@GeeAlmeida the short term solution is that we have to stop using the bot and do the posting the old fashioned way. You're going to need to adjust the text for the different platforms to make use of the usernames features.

We can revisit an automated solution in future if one becomes viable for our needs, but we have to put those needs first.

**Comment by ZoeLeBlanc** [comment]
Thanks to @jrladd for all his hard work on this! Just getting up to speed on the issue, and one thing is that I don't think the github you posted the issue to @GeeAlmeida is active. If you go to the actual Moa.party site https://moa.party/, they actually host their code through gitlab https://gitlab.com/fedstoa/moa, so we might need to open an issue here instead. Looking through their issues it seems like this an issue that others have brought up https://gitlab.com/fedstoa/moa/-/issues/40 and not sure it's been resolved. 

So one option is that we could take a look at their code base and try to add a fix PR for this issue, or find another service that allows us to cross post. Any preferences @jrladd ? I was looking for alternative cross posting services and didn't find a ton of options via Google, but am very new to this so happy to follow your lead.

**Comment by jrladd** [comment]
Thanks @ZoeLeBlanc. It does seem like Moa is working on this, as [this PR](https://gitlab.com/fedstoa/moa/-/merge_requests/10) I linked before suggests. Before looking for a new crossposter, I think our first step should be to see if a slight adjustment to the way we format usernames would solve this.

@GeeAlmeida could you send a test post from the Mastodon account that has a username in the `@user@twitter.com` format? Then if you could check how it shows up on Twitter, we'd know whether making that adjustment would solve the problem. You can use my Twitter username, `@johnrladd`, if you're worried about bugging anyone with a false ping.

**Comment by rivaquiroga** [comment]
@GeeAlmeida, let’s hope that the fix @jrladd is proposing works :crossed_fingers:. If not, maybe while we wait for Moa to fix the bug we can go back to your previous workflow for a while (that everything you post to Twitter shows up in Mastodon).

> Is there a way we could prepare two variations of each post, one including Mastodon handles (where available) and one including Twitter handles, then program the bot to select which to post where?

@anisa-hawes has a very good point here. Even if the person has the same handle on Twitter and Mastodon, if they are not on the same server, the mention will not work (e.g., [this toot](https://hcommons.social/@proghist/109762297131959253
)). Silvia uses the same handle on both platforms, but she is on fedihum.org, so the mention is not pointing to her profile. 
Regarding the work for editors, it will involve writing separate posts for each platform, right? And using the proper handle in each case. And if they don’t use one or both of the platforms, we use their full name (as we currently do when they are not on Twitter). It seems manageable.

**Comment by GeeAlmeida** [comment]
> Thanks @ZoeLeBlanc. It does seem like Moa is working on this, as [this PR](https://gitlab.com/fedstoa/moa/-/merge_requests/10) I linked before suggests. Before looking for a new crossposter, I think our first step should be to see if a slight adjustment to the way we format usernames would solve this.
> 
> @GeeAlmeida could you send a test post from the Mastodon account that has a username in the `@user@twitter.com` format? Then if you could check how it shows up on Twitter, we'd know whether making that adjustment would solve the problem. You can use my Twitter username, `@johnrladd`, if you're worried about bugging anyone with a false ping.

I tried @jrladd John's suggestion and... it did not work. The Mastodon post reads "At @ProgHist@twitter.com we believe that peer-review ..." and the Twitter repost reads "At ProgHist we believe that peer-review ..." (without the @).

I have now revoked Moa Party's permissions. So no more crossposting for now.

Should I stop the bot as well? It is currently posting to Mastodon, but with Twitter handles.

Next week I can try to search all the Twitter handles and we can then have a Mastodon bot that posts with Mastodon handles (or names as @rivaquiroga said).

**Comment by GeeAlmeida** [comment]
> Thanks to @jrladd for all his hard work on this! Just getting up to speed on the issue, and one thing is that I don't think the github you posted the issue to @GeeAlmeida is active. If you go to the actual Moa.party site https://moa.party/, they actually host their code through gitlab https://gitlab.com/fedstoa/moa, so we might need to open an issue here instead. Looking through their issues it seems like this an issue that others have brought up https://gitlab.com/fedstoa/moa/-/issues/40 and not sure it's been resolved.
> 
> So one option is that we could take a look at their code base and try to add a fix PR for this issue, or find another service that allows us to cross post. Any preferences @jrladd ? I was looking for alternative cross posting services and didn't find a ton of options via Google, but am very new to this so happy to follow your lead.

Thanks for spotting that Zoe @ZoeLeBlanc I was going to have to wait a loooong time for their reply! ;)

**Comment by GeeAlmeida** [comment]
So I have started to manually try to duplicate our publications in both social media platforms today, and I have a question to whoever knows better: is it ok to copy someone's tweet from Twitter and paste/publish into Mastodon? Most organisations/people are not yet on Mastodon. If they were I could just boost (ie. retweet) their publication.
I have just done this (copy from T, paste into M) and it feels a bit wrong.

Alternatively I could just retweet on Twitter and boost on Mastodon. Any thoughts on this?

**Comment by josircg** [comment]
Fresh news: Twitter is going to charge to use the API. I think the Moa service will cease to work (or it will charge us to use it). https://twitter.com/TwitterDev/status/1621026986784337922

**Comment by hawc2** [comment]
What if we just started posting one tweet/toot a week and did it manually?
Does the bot and frequent posting really increase engagement with our
lessons?

Especially when the platforms are in flux, automating this process seems
like more effort than its worth. Frequent posting by organizational
accounts might also hurt engagement imo

**Comment by GeeAlmeida** [comment]
Hi Alex @hawc2  - what do you mean by "Frequent posting by organizational accounts " - are you talking about the bot? Or retweets? Or what?

**Comment by GeeAlmeida** [comment]
@hawc2 On the point you raise about engagement, I agree that the bot does not generate engagement, no. But it does keep the page active. I am only here two days a week. The bot (when working fine, some time ago) would publish two times a week, on days I am not working. That made the profile active - there was always something new.

That is what I like about the bot.
There is always the aspect that Riva mentioned above, which is that (especially authors on Twitter) coulr see their work being talked about and would often retweet.

**Comment by hawc2** [comment]
Hey Gisele, I guess I thought the point of the bot was to post more
frequently than would be worth doing manually ... I don't really know
what's a normal number for journals to tweet, but PH seems to post more
than most accounts I follow. It might not cause much of a problem if we
just tweet less and try to make each tweet really count

**Comment by ZoeLeBlanc** [comment]
We might be able to check the bot's engagement by looking at google analytics so it wouldn't necessarily be impossible to answer your question @hawc2 .

For now I think we need to stop programmatically posting to Twitter at the very least (thanks for turning off Moa Party Gisele!). I believe the original impetus for automated posting was that it took some of the burden off Editors since it was setup before we had @GeeAlmeida on the team. We could revisit that calculus, but I do think automatic posting for Mastodon is still a worthwhile approach (not to mention John did all the work already to get the new bot setup). It's also worth noting we don't post on every social media platform (Facebook, Instagram, etc...) but rather where we think our audience will be, and I'm increasingly convinced that for the immediate future that will be Mastodon but open to other suggestions too

**Comment by GeeAlmeida** [comment]
Thanks @hawc2 I get you mean now.

I had a chat with Anisa today and we both think that the bot needs to be disabled for now, because it posts to Mastodon using Twitter handles. Next week I will try to search for all the handles in the bot sheet and turn them into Mastodon handles (when available) or full name (when not). This way we will have a perfectly operational bot for Mastodon, which is great. But unfortunately I am switching it until we have collected this intel.

**Comment by GeeAlmeida** [comment]
> We might be able to check the bot's engagement by looking at google analytics so it wouldn't necessarily be impossible to answer your question @hawc2 .
> 
> For now I think we need to stop programmatically posting to Twitter at the very least (thanks for turning off Moa Party Gisele!). I believe the original impetus for automated posting was that it took some of the burden off Editors since it was setup before we had @GeeAlmeida on the team. We could revisit that calculus, but I do think automatic posting for Mastodon is still a worthwhile approach (not to mention John did all the work already to get the new bot setup). It's also worth noting we don't post on every social media platform (Facebook, Instagram, etc...) but rather where we think our audience will be, and I'm increasingly convinced that for the immediate future that will be Mastodon but open to other suggestions too
[ProgHist Tweets Oct 22.csv](https://github.com/programminghistorian/jekyll/files/10580768/ProgHist.Tweets.Oct.22.csv)


I have extracted a month of tweets from TwitterAnalytics. It's October 2022.

**Comment by rivaquiroga** [comment]
@GeeAlmeida, you will still be able to schedule tweets using [Tweetdeck](https://tweetdeck.twitter.com/
). That will help keep the account active when you are not around. 

As Zoe says, increasingly more Twitter users will be moving to Mastodon. So it is important to take that into account. But we might also consider what is going on on other platforms. For example, many Latin American DH labs are active on Instagram (e.g. [Laboratório de Humanidades Digitais da UFBA](https://www.instagram.com/labhdufba/), [Laboratório de Humanidades Digitais PUC-Rio](https://www.instagram.com/labhdpucrio/), [Laboratorio de Humanidades Digitales PUCP](https://www.instagram.com/humanidadesdigitalespucp/), just to mention a few). One of the reasons is that grad and especially undergrad students down here are not usually Twitter/Mastodon users. So Instagram is the way to reach that audience.

**Comment by jrladd** [comment]
Thanks for this discussion, everyone! I saw yesterday that Twitter is disabling the API, which as others have mentioned really changes this whole conversation. (That's providing that Twitter actually *goes through* with this decision, which remains to be seen.) @ZoeLeBlanc and I plan to discuss this a bit more at the tech team meeting this week.

No access to the Twitter API will mean that neither Moa nor any kind of Twitter bot will work: all posting on Twitter will have to be done manually or scheduled with a Twitter-controlled tool like Tweetdeck. But Mastodon posting can still be automated as normal, and I think it's worthwhile keeping some automated posts going there. And as @hawc2 mentions we can examine the posting frequency and see what works for us: right now it posts once a day. The only other thing we would need to do is change the usernames on our spreadsheet from Twitter usernames to Mastodon ones. Since the bot will only be able to post in one place, that nicely simplifies things.

I agree with @rivaquiroga that we should think about diversifying our social media presence broadly, beyond just one or two sites, since different communities are migrating to different platforms. This is especially crucial as Twitter seems to be getting harder to work with as a central clearinghouse.

@GeeAlmeida @anisa-hawes I've temporarily paused the bot for now, until the tech team has a chance to talk about next steps at the meeting.

**Comment by anisa-hawes** [comment]
Thank you, @GeeAlmeida. I've made some edits to the English posts in the Google Sheet you've prepared.

An adjustment I’d like you to make across all languages, is to ensure that hashtags are written in the form known as “CamelCase”, where each new word is capitalised: #DiamondOpenAccess, #HumanidadesDigitais. This is an accessibility best-practice recommendation which makes it possible for screen reader technology to distinguish one word from another.

--

For now, I feel it would be most valuable to dedicate ~10-15 minutes on each of your two working days to publishing + scheduling one or two human-written posts on Twitter (including contributor user handles where available, full names where unavailable), and one or two human-written posts on Mastodon (including contributor user handles where available, full names where unavailable). As Riva says, combining manual posts with scheduling will keep the account active – my sense is that posting on 4 out of 7 days would be enough (echoing Alex's question: how frequently do comparable organisations post?). 

The TwitterAnalytics data you've downloaded gives quite interesting insights into which posts generated 'engagements' during October: URL clicks, clicks on our User Profile, new follows etc. So you will be able to invest your time into creating the kinds of posts that prompted action and meaningful engagement.

**Comment by GeeAlmeida** [comment]
I have started the process of identifying which authors are in Mastodon. I will describe here my process and if anyone has an idea of how to make it easier, that would be great.

(1) I used this formula in Excel to retrieve the Twitter handles: =MID(A1,FIND("@",A1)+1,FIND(" ",A1,FIND("@",A1))-FIND("@",A1)-1)
Source: https://www.mrexcel.com/board/threads/extract-twitter-handles-only.885302/

(2) I used Debirdify (https://debirdify.pruvisto.org/) to identify which of the people are in Mastodon. You can just past the whole list and it will check in one go. Debirdify seems to only grab one handle per line though, so where there are more than one author one needs to double check.

(3) I collected the information from Debirdify in the same column (H), and in the same Excel sheet where the tweets are (it's a copy of the bot)

(4) I clicked on the link for lessons and copied the name of the authors who don't have a Mastodon account into the same column (G)


If anyone is interested in helping out with this, either by doing a tab or by making the workflow easier, the file I am working on is here: https://bit.ly/3RTmKBw

I have only done the English so far.

**Comment by spapastamkou** [comment]
@GeeAlmeida Not directly on your last question but I wanted to let you know that the twitter account "[at]fdlaramee" is no longer active and should be replaced by the name: François Dominic Laramée (or FD Laramée if easier) in the sheet for the Twitter bot. He was a former FR team member and I have no knowledge of him having another account on Mastodon. You will find him as author in the EN and PT tweets and as author or translator or editor in the FR tweets of the bot. 
I myself may figure as translator in the FR tweets of the bot under the twitter account "[at]s_papastamkou". My mastodon account is: "[at]s_papastamkou[at]piaille.fr".

**Comment by anisa-hawes** [comment]
Thank you, @GeeAlmeida. I've made some edits to the English posts in the Google Sheet you've prepared.

An adjustment I’d like you to make across all languages, is to ensure that hashtags are written in the form known as “CamelCase”, where each new word is capitalised: #DiamondOpenAccess, #HumanidadesDigitais. This is an accessibility best-practice recommendation which makes it possible for screen reader technology to distinguish one word from another.

--

For now, I feel it would be most valuable to dedicate ~10-15 minutes on each of your two working days to publishing + scheduling one or two human-written posts on Twitter (including contributor user handles where available, full names where unavailable), and one or two human-written posts on Mastodon (including contributor user handles where available, full names where unavailable). As Riva says, combining manual posts with scheduling will keep the account active – my sense is that posting on 4 out of 7 days would be enough (echoing Alex's question: how frequently do comparable organisations post?).

The TwitterAnalytics data you've downloaded give quite interesting insights into which posts generated 'engagements' during October: URL clicks, clicks on our User Profile, new follows etc. My sense is that Twitter's 'impressions' measure is increasingly useless, because this metric can represent someone merely scrolling past a post in their feed. It is important to invest your time into creating only the kinds of posts that prompt action and meaningful engagement.

**Comment by GeeAlmeida** [comment]
Hi @anisa-hawes
Thanks for looking through the English posts! I'll try to see what you have changed so I can do the same to the Twitter bot file.

On the social media, here are my thoughts:
The Mastodon bot is almost done now. I have followed the procedure I described above, and the posts will have contributor user handles and correct server if they are on Mastodon, full names where unavailable. I suggest turning on the Mastodon bot and have it post to Mastodon twice a week (Mondays and Thursdays). 

Meanwhile, I will use the Twitter bot sheet to schedule tweets for Mondays and Thursdays.

And on the days I am here, I will continue to post new messages on both platforms.

@rivaquiroga Riva has made good points about the posts with links to lessons being a good way to thank the authors and give them visibility.

**Comment by GeeAlmeida** [comment]
Can anyone tell me where is the Twitter bot Google spreadsheet hosted? Which Google Drive? & who owns it?

I need to edit it but I can only visualise it - I sent a request to edit today. 

I think it would be a good idea to place both bots together - and have the Twitter bot Google spreadsheet and the new Mastodon bot Google spreadsheet in the same folder / Google drive.

**Comment by jenniferisasi** [comment]
Hi @GeeAlmeida! I just sent you an invite to be an editor for the spreadsheet plus some more information. I believe @anisa-hawes might be the one with full access cos it says the PH email is the proprietor. But hope editor access helps you for now.

**Comment by GeeAlmeida** [comment]
@jrladd I have good news!

The transformation of the Twitter bot into a Mastodon bot is complete. The new Google spreadsheet is here: https://docs.google.com/spreadsheets/d/168wpCJxQqMvIfCyC5SYjp8pC40SlPEKa/edit#gid=1140870057

Can you please switch it on when you get a chance? But I think you should set it to publish only twice a week, on Mondays and Thursdays.

**Comment by anisa-hawes** [comment]
@jrladd and I met last week, so I now have a (somewhat) clearer understanding of the Bot’s script + set up (thank you for making the time to explain things, John – please let me know if my summary below is accurate, or if it is missing anything technical you feel is needed).

Key points:
- The code that powered our original Twitter bot has been adjusted to power a Mastodon bot instead.
- The Mastodon bot’s code includes an adjustment that serves the system to run via GitHub Actions rather than Heroku.
- _If_ we can continue to access Twitter’s API then it will be simple to create a new repo, copy the Mastodon bot code, and adjust it back again to work as a Twitter bot. (Also adding John’s tweaks so that the Twitter bot also runs via GitHub Actions).

Problem:
- The Mastodon bot is currently disabled due to the way it treats the links created by the @ symbols that precede Twitter account handles.
- The cross-posting service we trialled (Moa Party) and (anecdotally) others too, are not managing this well at the moment. Moa was actually removing the @ symbols / links from our Twitter posts completely.
-  Gisele invested effort into creating a new version of the Google Sheet which plots in Mastodon account handles where  available and replaces Twitter account handles with real names. So we could go ahead — using and maintaining this Mastodon-only sheet.

There is some thinking/work to do:
- _If_ we can continue to use Twitter and want to use cross-posting, we need to work through the account handles problem.
- By adding '@twitter.com' as a suffix on every account handle included on the original Google Sheet, these would link correctly to Twitter when featured in a Mastodon post
- What happens to Mastodon account links when featured in a Twitter post?
- By appending a short python script to the bot’s code, we can ensure that whenever a post is pulled from the Google Sheet _to post on Twitter_  the suffix '@twitter.com' is removed
- Would we need to maintain and update two Google Sheets?

Some further points to consider:
- Is cross-posting meaningful/effective? 
- How might the increased character-allowance on Mastodon enable us to communicate in a different way? (Should we consider sharing whole abstracts?, for example)
- How are the needs and behaviours of the Mastodon community different? (What engagement analytics are available there? What can we learn about how people act upon our posts?) 
- Would we instead like to take this opportunity to re-think things a bit ? 
- our posting frequency: both bots use a 'crontab system’ which John explains makes it quite simple to adjust the times + days of the week we want to our bot to post.
- our `message-1/message-2` structure: use a single message per lesson instead?

We have a well-established following on Twitter, and a growing number of people in our network are beginning to use Mastodon. But how useful do they find our communications? I thought the Twitter ‘engagement metrics’ were interesting, because they revealed so little active engagement — our posts don’t seem to be operating as direct calls to action: Read a lesson (old or new); Support the project; Join an event; Contribute as an author /translator / editor / reviewer.

--

Although this Issue is technical, I think it raises some broader questions about community-building which we could take forwards in a different forum (perhaps as a focused group discussion?). Is there interest in developing this conversation among the @programminghistorian/project-team?

**Comment by jrladd** [comment]
This is a great summary, @anisa-hawes, and I'd be glad to participate in these broader conversations.

I wanted to add one thing. In addition to creating a separate bot to post on Twitter based on this code, we could also adapt this code to post to both Mastodon and Twitter simultaneously. This would depend on how we decide to deal with usernames, and as always it will depend on whether Twitter continues to allow free API use.

**Comment by hawc2** [comment]
Picking up this thread, are we still using Twitter in any way? I think their transformation into X should be the nail in the coffin

**Comment by spapastamkou** [comment]
Just drop here the fact that Tweetdeck also is soon to become inaccessible short of a subscription of some 100 $ or so annually.

**Comment by anisa-hawes** [comment]
I am closing this Issue because we no longer require a Mastodon Bot to fulfil our communications objectives.

-- 

We have introduced a number of new communications workflows to meet our goals:

**1. New Lesson Announcements**
- Formulated using templates (agreed and translated in dialogue with the Managing Editors) 
- Prepared by Charlotte
- Posted manually upon publication on Twitter/X and Mastodon
- Include newly activated DOIs
- Include all contributors' profile handles according to the platform (where available, full names where unavailable)

**2. New Lesson Reminders (~21 days after publication)**
- Formulated using templates (agreed and translated in dialogue with the Managing Editors) 
- Prepared by Charlotte
- Scheduled for ~21 days after publication. (We schedule in-platform on Twitter/X, and we use a free, third-party tool called Buffer to schedule for Mastodon)
- Include DOIs
- Include all contributors' profile handles according to the platform (where available, full names where unavailable)

**3. Future promotion of past lessons**
- Provided by Managing Editors for each lesson at Phase 7 Publication
- Formulated as pairs of 'evergreen' messages (shareable at any time in the future)
- Part 1 + Part 2 of an evergreen pair are posted + scheduled by Charlotte each week in either EN, ES, FR or PT (4-week rotation)
- Evergreens are selected at random from whichever language set is featured that week (ensuring that we haven't posted the evergreen pair for a translation of the same lesson recently) 
- Activity is logged with an XY (X = Part 1 posted, Y = Part 2 scheduled)
- Include DOIs 
- Include author/translator profile handles according to the platform (where available, full names where unavailable)

**4. Themed posts (evergreen)**
- Devised by Charlotte and Anisa (we are building a refreshed template bank of themed posts to be agreed and translated in dialogue with the Managing Editors)
- Formulated as 'evergreen' messages (shareable at any time in the future)
- Scheduled in-platform on Twitter/X, scheduled using Buffer for Mastodon 
- Theme 1. IPPs: encouraging new members
- Theme 2. Donations: encouraging new Patreon members and individual donations 
- Theme 3. What makes PH different? our CC BY 4.0 License; DiamondOpenAccess model; our peer review system 

**5. Themed posts (time-specific)**
- Devised by Charlotte and Anisa (agreed and translated in dialogue with the Managing Editors where necessary)
- Theme 1. IPPs: welcoming new members; thanking renewals
- Theme 2. Donations: welcoming new Patreons; thanking continuing subscribers
- Theme 3. News: research, events, awards, Bulletin/Buletín/Bulletin/Boletim quarterly issue announcements
- Theme 4. Calls: lesson proposals, editorships, reviewers

---

I think these communications workflows successfully mix templates and automatic scheduling with our human touch. Charlotte now dedicates a small amount of time at the beginning and end of her working week to communications. This time includes: posting + scheduling for the coming week, checking notifications, responding to engagements and questions.

Charlotte and I coordinate and fully manage these workflows using co-owned spreadsheets saved in a Social Media Documentation folder on our Google Workspace. This documentation includes a username log which records the Twitter and Mastodon profile handles of the current Project Team, contributors and IPPs. This serves as a quick reference and saves time when drafting future posts.

--

We have recently established a [profile on LinkedIn](https://www.linkedin.com/company/prog-hist/), so we will be expanding our communications activity there too. This is a space where we can network with our supporters, contributors and collaborators. We will be sharing longer-form content, cross-posting from our site blog, and linking to content shared on other social channels where appropriate.

--

Please note that the previously-used Twitter Bot and Mastodon Bot worksheets are now deprecated: they included many gaps, errors, duplications and invalid usernames.

**Comment by rivaquiroga** [comment]
Thanks @anisa-hawes and @charlottejmc for these new communications workflows 🎉

**Comment by jenniferisasi** [comment]
Just read this and although closed, I want to join Riva in thanking @anisa-hawes and @charlottejmc for taking on this very important work onward!

**Comment by DanielAlvesLABDH** [comment]
Great work @anisa-hawes and @charlottejmc! Many thanks for this. Wishing you all Happy Holidays (Boas Festas)!
